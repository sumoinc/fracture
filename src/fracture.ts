import { Project } from 'projen';
import { AppSync } from './appsync/appsync';
import { FractureComponent } from './component';
import { Entity, EntityOptions } from './model';

export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   */
  gendir?: string;
  /**
   * Enable AppSync
   */
  appsync?: boolean;
}

export class Fracture extends FractureComponent {
  public readonly gendir: string;
  public readonly appsync: boolean;

  constructor(
    project: Project,
    namespace?: string,
    options: FractureOptions = {},
  ) {
    super(project, namespace ?? 'fracture' );

    this.gendir = options.gendir ?? project.outdir;
    this.appsync = options.appsync ?? true;

    if (this.appsync) {
      new AppSync(project, this.namespace, { gendir: this.gendir });
    }
  }

  /**
   * Get all entities in this project's namespace.
   */
  public get entities(): Entity[] {
    const isEntity = (c: FractureComponent): c is Entity => c instanceof Entity && c.namespace === this.namespace;
    return (this.project.components as FractureComponent[]).filter(isEntity);
  }

  public addEntity(options: EntityOptions) {
    return new Entity(this.project, this.namespace, options);
  }


}
