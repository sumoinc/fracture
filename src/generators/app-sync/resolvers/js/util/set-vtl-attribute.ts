import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
