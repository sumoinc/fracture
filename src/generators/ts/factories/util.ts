import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
