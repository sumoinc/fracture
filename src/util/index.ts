export * from "./determineBuildId";
export * from "./test-util";
