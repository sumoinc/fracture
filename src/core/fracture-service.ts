import { join } from "path";
import { paramCase } from "change-case";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { Fracture } from "./fracture";
import { Resource, ResourceOptions } from "./resource";
import { Structure, StructureOptions } from "./structure";
import { StructureAttributeType } from "./structure-attribute";
import { DynamoTable } from "../dynamodb";
import * as ts from "../generators/ts";
import {
  TypescriptStrategy,
  TypescriptStrategyOptions,
} from "../generators/ts/strategy";

export interface FractureServiceOptions
  extends Partial<TypeScriptProjectOptions> {
  /**
   * A name for the service
   */
  name: string;
  /**
   * Typescript strategy options
   *
   * @default fracture defaults
   */
  typescriptStrategyOptions?: TypescriptStrategyOptions;
  // srcDir?: string;
  /**
   * Logging options
   * @default LogLevel.INFO
   */
  //logging?: LoggerOptions;
  /**
   * Versioned.
   * @default false
   */
  //isVersioned?: boolean;
  /**
   * The naming strategy to use for generated code.
   */
  // namingStrategy?: NamingStrategy;
  /**
   * The audit strategy to use for generated code.
   */
  //auditStrategy?: AuditStrategy;
}

export class FractureService extends TypeScriptProject {
  /**
   * A name for the service.
   */
  public readonly name: string;
  /**
   * Defines typescript naming conventions for this service.
   */
  public readonly typescriptStrategy: TypescriptStrategy;
  /**
   * All resourcers in this service.
   */
  public resources: Resource[] = [];
  /**
   * Table defnition for the table that supports this service.
   */
  public readonly dynamoTable: DynamoTable;
  /**
   * All structures for this service.
   */
  public structures: Structure[] = [];
  public readonly errorStructure: Structure;
  public readonly responseStructure: Structure;
  public readonly listResponseStructure: Structure;

  constructor(fracture: Fracture, options: FractureServiceOptions) {
    /***************************************************************************
     * Projen Props
     **************************************************************************/

    // ensure name is param-cased for outdir
    const outdir = join(fracture.serviceRoot, paramCase(options.name));

    const projenOptions: TypeScriptProjectOptions = {
      name: options.name,
      defaultReleaseBranch: "main",
      parent: fracture,
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",
      outdir,
      eslintOptions: {
        dirs: ["src"],
        tsconfigPath: "./**/tsconfig.dev.json",
      },
      licensed: false,
      projenrcTs: true,
    };

    super(projenOptions);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.name = options.name;
    this.typescriptStrategy = new TypescriptStrategy(this);

    /***************************************************************************
     * Dynamo Configuration
     **************************************************************************/

    this.dynamoTable = new DynamoTable(this, { name: this.name });

    /***************************************************************************
     * Initialize standard structures
     **************************************************************************/

    this.errorStructure = this.addStructure({
      name: `error`,
      attributes: [
        {
          name: `code`,
          type: StructureAttributeType.INT,
        },
        {
          name: `source`,
        },
        {
          name: `message`,
        },
        {
          name: `detail`,
        },
      ],
    });

    this.responseStructure = this.addStructure({
      name: `response`,
      typeParameter: `T`,
      attributes: [
        {
          name: `data`,
          type: StructureAttributeType.CUSTOM,
          typeParameter: `T`,
          required: false,
        },
        {
          name: `errors`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `error`,
        },
        {
          name: `status`,
          type: StructureAttributeType.INT,
        },
      ],
    });

    this.listResponseStructure = this.addStructure({
      name: `list-response`,
      typeParameter: `T`,
      attributes: [
        {
          name: `data`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `T`,
          required: false,
        },
        {
          name: `errors`,
          type: StructureAttributeType.ARRAY,
          typeParameter: `error`,
        },
        {
          name: `status`,
          type: StructureAttributeType.INT,
        },
      ],
    });

    /***************************************************************************
     * Typescript Generators
     **************************************************************************/

    const typesFile = join("generated", "ts", "types.ts");

    new ts.TypescriptServiceTypes(this, typesFile);
  }

  public addResource(options: ResourceOptions) {
    const resource = new Resource(this, options);
    this.resources.push(resource);
    return resource;
  }

  public addStructure(options: StructureOptions) {
    const structure = new Structure(this, options);
    this.structures.push(structure);
    return structure;
  }

  //   /***************************************************************************
  //    *
  //    * CREATE PACKAGE SUB-PROJECT
  //    *
  //    * This powers a sub-project to house all generated code as a package. This
  //    * is mostly generated code and should not be touched by devs unless really
  //    * needed.
  //    *
  //    **************************************************************************/

  //   const project = new TypeScriptProject({
  //     defaultReleaseBranch: "main",
  //     name: this.packageName,
  //     parent: fracture,
  //     licensed: false,
  //     outdir: join(fracture.packageRoot, mergedOptions.name),
  //     packageManager: NodePackageManager.PNPM,
  //     pnpmVersion: "8",
  //     prettier: true,
  //     projenrcTs: true,
  //     deps: [
  //       "@aws-sdk/client-dynamodb",
  //       "@aws-sdk/lib-dynamodb",
  //       "aws-cdk",
  //       "aws-cdk-lib",
  //       "constructs",
  //       "uuid",
  //     ],

  //     devDeps: ["@types/uuid"],
  //     eslintOptions: {
  //       dirs: ["src"],
  //       tsconfigPath: "./**/tsconfig.dev.json",
  //     },
  //   });
  //   this.project = project;

  //   /***************************************************************************
  //    *
  //    * INIT SERVICE
  //    *
  //    **************************************************************************/

  //   this.fracture.logger.info("-".repeat(80));
  //   this.fracture.logger.info(`INIT Service: "${this.name}"`);
  //   this.fracture.logger.info("-".repeat(80));

  //   // inverse
  //   this.fracture.services.push(this);

  //   /***************************************************************************
  //    *
  //    * DYNAMO TABLE
  //    *
  //    * Add dynamo table per options defined in service definition
  //    *
  //    **************************************************************************/

  //   this.dynamoTable = new DynamoTable(this);

  //   // add Dynalite support for Jest tests
  //   this.dynaliteSupport = new DynaliteSupport(this);

  //   /***************************************************************************
  //    *
  //    * GENERATORS
  //    *
  //    **************************************************************************/

  //   //this.ts = new TypescriptService(this);

  //   this.tsTypes = new TypescriptTypes(this);
  //   this.cdkApp = new CdkApp(this);

  //   return this;
  // }

  // public get name(): string {
  //   return this.options.name;
  // }

  // public get packageName(): string {
  //   return `@${this.fracture.name}/${this.name}-pkg`;
  // }

  // public get serviceName(): string {
  //   return `@${this.fracture.name}/${this.name}-svc`;
  // }

  // public get isVersioned(): boolean {
  //   return this.options.isVersioned;
  // }

  // public get namingStrategy(): NamingStrategy {
  //   return this.options.namingStrategy;
  // }

  // public get auditStrategy() {
  //   return this.options.auditStrategy;
  // }

  // public get srcDir() {
  //   return this.options.srcDir;
  // }

  // public get serviceRoot() {
  //   return join(this.fracture.packageRoot, this.name);
  // }

  // /**
  //  * Returns index for this package in the overall project.
  //  * Useful when trying to split up ports for testing in parallel, etc.
  //  */
  // public get serviceIndex() {
  //   const { services } = this.fracture;
  //   return services.findIndex((p) => p.name === this.name) || 0;
  // }

  // /*****************************************************************************
  //  *
  //  *  Configuration Helpers
  //  *
  //  ****************************************************************************/

  // /**
  //  * Add a resource to the fracture project.
  //  *
  //  * @param {ResourceOptions}
  //  * @returns {Resource}
  //  */
  // public addResource(options: ResourceOptions) {
  //   return new Resource(this, options);
  // }

  // public get keyDynamoGsi(): DynamoGsi {
  //   return this.dynamoTable.keyDynamoGsi;
  // }

  // public get lookupDynamoGsi(): DynamoGsi {
  //   return this.dynamoTable.lookupDynamoGsi;
  // }

  // /*****************************************************************************
  //  *
  //  *  TYPESCRIPT HELPERS
  //  *
  //  ****************************************************************************/

  // public get tsResponseTypeName() {
  //   return formatStringByNamingStrategy(
  //     "response",
  //     this.namingStrategy.ts.typeName
  //   );
  // }

  // public get tsLlistResponseTypeName() {
  //   return formatStringByNamingStrategy(
  //     "list-response",
  //     this.namingStrategy.ts.typeName
  //   );
  // }
}
