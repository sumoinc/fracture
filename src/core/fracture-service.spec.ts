import { Fracture } from "./fracture";
import { FractureService } from "./fracture-service";
import { synthFile, synthFiles } from "../util/test-util";

let fracture: Fracture;
let service: FractureService;

beforeEach(() => {
  fracture = new Fracture();
  service = new FractureService(fracture, { name: "foo" });
});

test("Smoke test", () => {
  expect(service).toBeTruthy();
});
describe("validate generated project files", () => {
  test("Detect new files", () => {
    const content = synthFiles(fracture, "services");
    const fileList = Object.keys(content);
    expect(fileList).toMatchSnapshot();
    //console.log(JSON.stringify(fileList, null, 2));
  });

  test("services/foo/.eslintrc.json", () => {
    const content = synthFile(fracture, "services/foo/.eslintrc.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.gitattributes", () => {
    const content = synthFile(fracture, "services/foo/.gitattributes");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.gitignore", () => {
    const content = synthFile(fracture, "services/foo/.gitignore");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.npmignore", () => {
    const content = synthFile(fracture, "services/foo/.npmignore");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.npmrc", () => {
    const content = synthFile(fracture, "services/foo/.npmrc");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.projen/deps.json", () => {
    const content = synthFile(fracture, "services/foo/.projen/deps.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.projen/files.json", () => {
    const content = synthFile(fracture, "services/foo/.projen/files.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/.projen/tasks.json", () => {
    const content = synthFile(fracture, "services/foo/.projen/tasks.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/package.json", () => {
    const content = synthFile(fracture, "services/foo/package.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/README.md", () => {
    const content = synthFile(fracture, "services/foo/README.md");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/src/index.ts", () => {
    const content = synthFile(fracture, "services/foo/src/index.ts");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/test/hello.test.ts", () => {
    const content = synthFile(fracture, "services/foo/test/hello.test.ts");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/tsconfig.dev.json", () => {
    const content = synthFile(fracture, "services/foo/tsconfig.dev.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
  test("services/foo/tsconfig.json", () => {
    const content = synthFile(fracture, "services/foo/tsconfig.json");
    expect(content).toMatchSnapshot();
    // console.log(JSON.stringify(content, null, 2));
  });
});
