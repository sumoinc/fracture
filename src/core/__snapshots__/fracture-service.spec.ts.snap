// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate generated project files Detect new files 1`] = `
Array [
  "services/foo/.eslintrc.json",
  "services/foo/.gitattributes",
  "services/foo/.gitignore",
  "services/foo/.npmignore",
  "services/foo/.npmrc",
  "services/foo/.projen/deps.json",
  "services/foo/.projen/files.json",
  "services/foo/.projen/tasks.json",
  "services/foo/generated/ts/types.ts",
  "services/foo/package.json",
  "services/foo/README.md",
  "services/foo/src/index.ts",
  "services/foo/test/hello.test.ts",
  "services/foo/tsconfig.dev.json",
  "services/foo/tsconfig.json",
]
`;

exports[`validate generated project files services/foo/.eslintrc.json 1`] = `
Object {
  "env": Object {
    "jest": true,
    "node": true,
  },
  "extends": Array [
    "plugin:import/typescript",
  ],
  "ignorePatterns": Array [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
  ],
  "overrides": Array [],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "ecmaVersion": 2018,
    "project": "./**/tsconfig.dev.json",
    "sourceType": "module",
  },
  "plugins": Array [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": Object {
    "@typescript-eslint/indent": Array [
      "error",
      2,
    ],
    "@typescript-eslint/member-delimiter-style": Array [
      "error",
    ],
    "@typescript-eslint/member-ordering": Array [
      "error",
      Object {
        "default": Array [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": Array [
      "error",
    ],
    "@typescript-eslint/no-require-imports": Array [
      "error",
    ],
    "@typescript-eslint/no-shadow": Array [
      "error",
    ],
    "@typescript-eslint/return-await": Array [
      "error",
    ],
    "array-bracket-newline": Array [
      "error",
      "consistent",
    ],
    "array-bracket-spacing": Array [
      "error",
      "never",
    ],
    "brace-style": Array [
      "error",
      "1tbs",
      Object {
        "allowSingleLine": true,
      },
    ],
    "comma-dangle": Array [
      "error",
      "always-multiline",
    ],
    "comma-spacing": Array [
      "error",
      Object {
        "after": true,
        "before": false,
      },
    ],
    "curly": Array [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": Array [
      "error",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": Array [
          "**/test/**",
          "**/build-tools/**",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": Array [
      "error",
    ],
    "import/order": Array [
      "warn",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": Array [
          "builtin",
          "external",
        ],
      },
    ],
    "indent": Array [
      "off",
    ],
    "key-spacing": Array [
      "error",
    ],
    "keyword-spacing": Array [
      "error",
    ],
    "max-len": Array [
      "error",
      Object {
        "code": 150,
        "ignoreComments": true,
        "ignoreRegExpLiterals": true,
        "ignoreStrings": true,
        "ignoreTemplateLiterals": true,
        "ignoreUrls": true,
      },
    ],
    "no-bitwise": Array [
      "error",
    ],
    "no-duplicate-imports": Array [
      "error",
    ],
    "no-multi-spaces": Array [
      "error",
      Object {
        "ignoreEOLComments": false,
      },
    ],
    "no-multiple-empty-lines": Array [
      "error",
    ],
    "no-return-await": Array [
      "off",
    ],
    "no-shadow": Array [
      "off",
    ],
    "no-trailing-spaces": Array [
      "error",
    ],
    "object-curly-newline": Array [
      "error",
      Object {
        "consistent": true,
        "multiline": true,
      },
    ],
    "object-curly-spacing": Array [
      "error",
      "always",
    ],
    "object-property-newline": Array [
      "error",
      Object {
        "allowAllPropertiesOnSameLine": true,
      },
    ],
    "quote-props": Array [
      "error",
      "consistent-as-needed",
    ],
    "quotes": Array [
      "error",
      "single",
      Object {
        "avoidEscape": true,
      },
    ],
    "semi": Array [
      "error",
      "always",
    ],
    "space-before-blocks": Array [
      "error",
    ],
  },
  "settings": Object {
    "import/parsers": Object {
      "@typescript-eslint/parser": Array [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": Object {
      "node": Object {},
      "typescript": Object {
        "alwaysTryTypes": true,
        "project": "./**/tsconfig.dev.json",
      },
    },
  },
}
`;

exports[`validate generated project files services/foo/.gitattributes 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.eslintrc.json linguist-generated
/.gitattributes linguist-generated
/.gitignore linguist-generated
/.npmignore linguist-generated
/.npmrc linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/generated/ts/types.ts linguist-generated
/package.json linguist-generated
/pnpm-lock.yaml linguist-generated
/tsconfig.dev.json linguist-generated
/tsconfig.json linguist-generated"
`;

exports[`validate generated project files services/foo/.gitignore 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/package.json
!/.npmignore
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
pids
*.pid
*.seed
*.pid.lock
lib-cov
coverage
*.lcov
.nyc_output
build/Release
node_modules/
jspm_packages/
*.tsbuildinfo
.eslintcache
*.tgz
.yarn-integrity
.cache
!/.projenrc.js
/test-reports/
junit.xml
/coverage/
!/.npmrc
!/test/
!/tsconfig.json
!/tsconfig.dev.json
!/src/
/lib
/dist/
!/.eslintrc.json
!/generated/ts/types.ts
"
`;

exports[`validate generated project files services/foo/.npmignore 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
/.projen/
/test-reports/
junit.xml
/coverage/
/test/
/tsconfig.dev.json
/src/
!/lib/
!/lib/**/*.js
!/lib/**/*.d.ts
dist
/tsconfig.json
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.eslintrc.json
"
`;

exports[`validate generated project files services/foo/.npmrc 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

resolution-mode=highest
"
`;

exports[`validate generated project files services/foo/.projen/deps.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "dependencies": Array [
    Object {
      "name": "@types/jest",
      "type": "build",
    },
    Object {
      "name": "@types/node",
      "type": "build",
      "version": "^16",
    },
    Object {
      "name": "@typescript-eslint/eslint-plugin",
      "type": "build",
      "version": "^5",
    },
    Object {
      "name": "@typescript-eslint/parser",
      "type": "build",
      "version": "^5",
    },
    Object {
      "name": "eslint-import-resolver-node",
      "type": "build",
    },
    Object {
      "name": "eslint-import-resolver-typescript",
      "type": "build",
    },
    Object {
      "name": "eslint-plugin-import",
      "type": "build",
    },
    Object {
      "name": "eslint",
      "type": "build",
      "version": "^8",
    },
    Object {
      "name": "jest",
      "type": "build",
    },
    Object {
      "name": "jest-junit",
      "type": "build",
      "version": "^15",
    },
    Object {
      "name": "npm-check-updates",
      "type": "build",
      "version": "^16",
    },
    Object {
      "name": "projen",
      "type": "build",
    },
    Object {
      "name": "ts-jest",
      "type": "build",
    },
    Object {
      "name": "typescript",
      "type": "build",
    },
  ],
}
`;

exports[`validate generated project files services/foo/.projen/files.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "files": Array [
    ".eslintrc.json",
    ".gitattributes",
    ".gitignore",
    ".npmignore",
    ".npmrc",
    ".projen/deps.json",
    ".projen/files.json",
    ".projen/tasks.json",
    "tsconfig.dev.json",
    "tsconfig.json",
  ],
}
`;

exports[`validate generated project files services/foo/.projen/tasks.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
      "steps": Array [
        Object {
          "cwd": "../..",
          "exec": "npx projen default",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools",
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
      "steps": Array [
        Object {
          "exec": "mkdir -p dist/js",
        },
        Object {
          "exec": "mv $(npm pack) dist/js/",
        },
      ],
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
        Object {
          "spawn": "eslint",
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpm update npm-check-updates",
        },
        Object {
          "exec": "npm-check-updates --upgrade --target=minor --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,eslint-import-resolver-node,eslint-import-resolver-typescript,eslint-plugin-import,eslint,jest,jest-junit,npm-check-updates,projen,ts-jest,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser eslint-import-resolver-node eslint-import-resolver-typescript eslint-plugin-import eslint jest jest-junit npm-check-updates projen ts-jest typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;

exports[`validate generated project files services/foo/README.md 1`] = `"# replace this"`;

exports[`validate generated project files services/foo/package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "devDependencies": Object {
    "@types/jest": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "eslint": "^8",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "jest": "*",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "projen": "*",
    "ts-jest": "*",
    "typescript": "*",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "name": "foo",
  "pnpm": Object {},
  "scripts": Object {
    "build": "npx projen build",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;

exports[`validate generated project files services/foo/src/index.ts 1`] = `
"export class Hello {
  public sayHello() {
    return 'hello, world!';
  }
}"
`;

exports[`validate generated project files services/foo/test/hello.test.ts 1`] = `
"import { Hello } from '../src';

test('hello', () => {
  expect(new Hello().sayHello()).toBe('hello, world!');
});"
`;

exports[`validate generated project files services/foo/tsconfig.dev.json 1`] = `
Object {
  "compilerOptions": Object {
    "alwaysStrict": true,
    "declaration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "lib": Array [
      "es2019",
    ],
    "module": "CommonJS",
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "stripInternal": true,
    "target": "ES2019",
  },
  "exclude": Array [
    "node_modules",
  ],
  "include": Array [
    ".projenrc.js",
    "src/**/*.ts",
    "test/**/*.ts",
  ],
}
`;

exports[`validate generated project files services/foo/tsconfig.json 1`] = `
Object {
  "compilerOptions": Object {
    "alwaysStrict": true,
    "declaration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "lib": Array [
      "es2019",
    ],
    "module": "CommonJS",
    "noEmitOnError": false,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "lib",
    "resolveJsonModule": true,
    "rootDir": "src",
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "stripInternal": true,
    "target": "ES2019",
  },
  "exclude": Array [],
  "include": Array [
    "src/**/*.ts",
  ],
}
`;
